### PR TITLE
Add do-rewrite construction

### PR DESCRIPTION
**Pull request by be5invis**
A fix for #3840.

Syntax:

```idris
do { rewrite H; ... }
```

Definition

```idris
do { rewrite H; ... } =====> rewrite H in do {...}
```

### CODE REVIEW

**Comment by Melvar** [comment]
I think we should remove `rewrite` instead of letting it worm its way into more parts of the language.

**Comment by be5invis** [comment]
@Melvar Without `rewrite` some proofs using `replace` would become extremely hard to write.

**Comment by be5invis** [comment]
cc. @ahmadsalim

**Comment by melted** [comment]
This has to be looked at by @edwinb before it can go in.

**Comment by ahmadsalim** [comment]
@be5invis Thanks for the PR!
Could you please add some tests to show that it works, and that it does not break existing syntax with `rewrite`-`in`? Could you also please update the `CHANGELOG` with proposed updates.

@Melvar I think in general that `rewrite` is a fairly useful language feature. Did @edwinb hint at wanting to remove it?

@melted I think this change is fairly uncontroversial, since all it is doing is adding an inline `rewrite` in `do` similarly to how we have inline `let` in `do`.
The only potential thing to worry about is backwards compatibility.

File: src/Idris/IBC.hs [hunk 1]
```diff
@@ -2292,6 +2292,9 @@ instance (Binary t) => Binary (PDo' t) where
                                       put x1
                                       put x2
                                       put x3
+                DoRewrite x1 x2 -> do putWord8 5
```
Comment (ziman) [inline src/Idris/IBC.hs]: Do we need to bump `ibcVersion` on line 50 of this file?

Comment (melted) [inline src/Idris/IBC.hs]: Good catch, that is definitely needed.

**Comment by david-christiansen** [comment]
I think this is a very nice feature, but given how central it is, we should let @edwinb weigh in.

**Comment by edwinb** [comment]
Looks fine to me. It's consistent with the way 'let' syntax works, too. Sorry for not paying attention to these...

@Melvar what do you see as the problem with `rewrite`? I find it invaluable for programs which do fiddly things (or even fairly mundane things) with indices, e.g. https://github.com/edwinb/Blodwen/blob/master/src/Core/TT.idr#L438 or https://github.com/edwinb/Blodwen/blob/master/src/Core/Evaluate.idr#L137

**Comment by Melvar** [comment]
I’m guessing I wrote that comment on a bad day; really all it is is that I have a pet peeve about `rewrite … in Refl`, always having to mentally convert it into `cong (…)` to figure out what it is doing.

**Comment by melted** [comment]
@be5invis could you run stylize on it? Then we can merge.

**Pull request merged**